cmd/lotus-shed/migrations.go

```diff
@@ -171,58 +213,361 @@ func checkStateInvariants(ctx context.Context, oldStateRoot cid.Cid, newStateRoo
 		return nil, err
 	}
 
-	verifregV8, err := stateTreeV8.GetActor(verifreg.Address)
+	return verifregDatacaps, nil
+}
+
+func getDatacap9Datacaps(stateTreeV9 *state.StateTree, actorStore adt.Store) (map[address.Address]abi.StoragePower, error) {
+	datacapStateV9, err := getDatacapActorV9(stateTreeV9, actorStore)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get datacap actor state: %w", err)
+	}
+
+	var datacaps = make(map[address.Address]abi.StoragePower)
+	err = datacapStateV9.ForEachClient(func(addr address.Address, dcap abi.StoragePower) error {
+		datacaps[addr] = dcap
+		return nil
+	})
 	if err != nil {
 		return nil, err
 	}
 
-	verifregV8State, err := verifreg.Load(actorStore, verifregV8)
-	if err = actorStore.Get(ctx, verifregV8.Head, &verifregV8State); err != nil {
-		return nil, xerrors.Errorf("failed to get verifreg actor state: %w", err)
+	return datacaps, nil
+}
+
+func checkPendingVerifiedDeals(stateTreeV8 *state.StateTree, stateTreeV9 *state.StateTree, actorStore adt.Store) error {
+	marketActorV9, err := getMarketActorV9(stateTreeV9, actorStore)
+	if err != nil {
+		return err
 	}
 
-	var verifregDatacaps = make(map[address.Address]abi.StoragePower)
-	err = verifregV8State.ForEachClient(func(addr address.Address, dcap abi.StoragePower) error {
-		verifregDatacaps[addr] = dcap
+	verifregActorV9, err := getVerifregActorV9(stateTreeV9, actorStore)
+	if err != nil {
+		return err
+	}
+
+	verifregStateV9, err := getVerifregStateV9(stateTreeV9, actorStore)
+	if err != nil {
+		return err
+	}
+
+	marketStateV8, err := getMarketStateV8(stateTreeV8, actorStore)
+	if err != nil {
+		return err
+	}
+
+	marketStateV9, err := getMarketStateV9(stateTreeV9, actorStore)
+	if err != nil {
+		return err
+	}
+
+	pendingProposalsV8, err := adt8.AsSet(actorStore, marketStateV8.PendingProposals, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return xerrors.Errorf("failed to load pending proposals: %w", err)
+	}
+
+	dealProposalsV8, err := market8.AsDealProposalArray(actorStore, marketStateV8.Proposals)
+	if err != nil {
+		return xerrors.Errorf("failed to get proposals: %w", err)
+	}
+
+	var numPendingVerifiedDeals = 0
+	var proposal market8.DealProposal
+	err = dealProposalsV8.ForEach(&proposal, func(dealID int64) error {
+		// If not verified, do nothing
+		if !proposal.VerifiedDeal {
+			return nil
+		}
+
+		pcid, err := proposal.Cid()
+		if err != nil {
+			return err
+		}
+
+		isPending, err := pendingProposalsV8.Has(abi.CidKey(pcid))
+		if err != nil {
+			return xerrors.Errorf("failed to check pending: %w", err)
+		}
+
+		// Nothing to do for not-pending deals
+		if !isPending {
+			return nil
+		}
+
+		numPendingVerifiedDeals++
+		// Checks if allocation ID is in market map
+		allocationId, err := marketActorV9.GetAllocationIdForPendingDeal(abi.DealID(dealID))
+		if err != nil {
+			return err
+		}
+		// Checks if allocation is in verifreg
+		_, found, err := verifregActorV9.GetAllocation(proposal.Client, allocationId)
```

Comment:
I would assert that the Allocation is as expected here (Client, provider, piece cid, etc)